app/src/main/java/org/dhis2/usescases/datasets/dataSetTable/dataSetSection/DataValuePresenter.kt

```diff
@@ -0,0 +1,741 @@
+                        ArrayList(),
+                        "android",
+                        row,
+                        column,
+                        categoryOptionCombo.uid(),
+                        dataTableModel.catCombo()!!.uid()
+                    )
+                }
+
+                fields.add(fieldViewModel)
+                values.add(fieldViewModel.value().toString())
+
+                if (!section!!.uid().isEmpty() && section!!.showRowTotals()!! &&
+                    isNumber && !fieldViewModel.value()!!.isEmpty()
+                ) {
+                    totalRow += Integer.parseInt(fieldViewModel.value()!!)
+                }
+
+                column++
+            }
+
+            for (fieldViewModel in fields)
+                for (compulsoryDataElement in dataTableModel.compulsoryCells()!!)
+                    if (compulsoryDataElement.categoryOptionCombo()!!.uid() ==
+                        fieldViewModel.categoryOptionCombo() &&
+                        compulsoryDataElement.dataElement()!!.uid() == fieldViewModel.dataElement()
+                    ) {
+                        fields[fields.indexOf(fieldViewModel)] = fieldViewModel.setMandatory()
+                    }
+
+            if (!section!!.uid().isEmpty() && section!!.showRowTotals()!! && isNumber) {
+                setTotalRow(totalRow, fields, values, row, column)
+            }
+
+            listFields.add(fields)
+            cells.add(values)
+            column = 0
+            row++
+        }
+
+        tableCells.add(listFields)
+
+        if (isNumber) {
+            if (section!!.uid().isNotEmpty() && section!!.showColumnTotals()!!) {
+                setTotalColumn(listFields, cells, dataTableModel.rows()!!, row, column)
+            }
+            if (section!!.uid().isNotEmpty() && section!!.showRowTotals()!!) {
+                for (i in 0 until dataTableModel.header()!!.size) {
+                    if (i == dataTableModel.header()!!.size - 1) {
+                        dataTableModel.header()!![i].add(
+                            CategoryOption.builder().uid("").displayName(
+                                "Total"
+                            ).build()
+                        )
+                    } else {
+                        dataTableModel.header()!![i].add(
+                            CategoryOption.builder().uid("").displayName(
+                                ""
+                            ).build()
+                        )
+                    }
+                }
+            }
+        }
+
+        val isEditable = accessDataWrite &&
+            !isExpired(dataSet) &&
+            dataInputPeriodModel.isEmpty() || (
+            checkHasInputPeriod() != null && DateUtils.getInstance().isInsideInputPeriod(
+                checkHasInputPeriod()
+            )
+            ) &&
+            !isApproval
+
+        return Quartet.create(dataTableModel, listFields, cells, isEditable)
```

Comment:
Not for now. But this method is way too long (155 lines).
Would be better to split it into several smaller methods.

Reply:
or classes I see too many responsabilities in same method and things that perhaps the presenter should not do. Let's take into account next time when performing unit tests after the release.